6.1

NikitaBabich26rus opened the pull request.

NikitaBabich26rus committed:
6.1

yurii-litvinov commented on HomeWork6/6.1/6.1/6.1/Stack.cpp:
```diff
@@ -0,0 +1,19 @@
+
+#include "Stack.h"
+
+using namespace std;
```
using namespace std; нельзя

yurii-litvinov commented on HomeWork6/6.1/6.1/6.1/Stack.h:
```diff
@@ -0,0 +1,11 @@
+	Stack* next;
+};
+
+void push(int value, Stack** head);
```
Надо комментарии ко всем функциям и структурам в заголовочном файле

yurii-litvinov commented on HomeWork6/6.1/6.1/6.1/6.1.cpp:
```diff
@@ -0,0 +1,99 @@
+
+int main()
+{
+	printf("%d", tests());
```
Это никак не влияет на код возврата и делает невозможным узнать о результате тестирования из сторонних программ или скриптов

yurii-litvinov commented on HomeWork6/6.1/6.1/6.1/6.1.cpp:
```diff
@@ -0,0 +1,99 @@
+int tests()
+{
+	int checkTest = 0;
+	Stack* headTest1 = nullptr;
```
Не удаляется, утечка памяти

yurii-litvinov commented on HomeWork6/6.1/6.1/6.1/6.1.cpp:
```diff
@@ -0,0 +1,99 @@
+	checkString(&headTest2, stringTest2, sizeTest2);
+	printf("My answer : ");
+	printf("%d\n", headTest2->value);
+	checkTest = checkAnswer(headTest2->value, answerTest2, checkTest);
```
А если первый тест не прошёл, а второй прошёл, что будет в checkTest?

yurii-litvinov commented on HomeWork6/6.1/6.1/6.1/6.1.cpp:
```diff
@@ -0,0 +1,99 @@
+	}
+}
+
+void checkString(Stack** head, char string[], int size)
```
Думаю, size всегда равен strlen(string), поэтому его можно не передавать. Кроме того, у строки в конце символ `\0`, так что обычно делают цикл не `for (int i = 0; i < size; i++)`, а `for (int i = 0; string[i] != '\0'; i++)`

yurii-litvinov commented on HomeWork6/6.1/6.1/6.1/6.1.cpp:
```diff
@@ -0,0 +1,99 @@
+
+void operation(Stack **head, char operation)
+{
+	const int value1 = (*head)->next->value;
```
Эх, это хорошо если пользовательский ввод корректен. А как это будет работать на тесте `1 +`?

NikitaBabich26rus committed:
6.1 // исправленно

yurii-litvinov commented on HomeWork6/6.1/6.1/6.1/6.1.cpp:
```diff
@@ -0,0 +1,132 @@
+bool testCorrect(char string[])
+{
+	int amountValues = 0;
+	for (int i = 0; i < strlen(string); i++)
```
strlen на каждой итерации цикла будет считать длину строки заново. Правильнее `for (int i = 0; string[i] != '\0'; ++i)`

yurii-litvinov commented on HomeWork6/6.1/6.1/6.1/6.1.cpp:
```diff
@@ -0,0 +1,132 @@
+	}
+}
+
+int testing(Stack** headTest, char stringTest[], int checkTest, int answerTest)
```
А чего testing не может себе  стек создать сам? Он ведь его удаляет, было бы логично и создавать. И это сэкономило бы параметры функции

yurii-litvinov commented on HomeWork6/6.1/6.1/6.1/6.1.cpp:
```diff
@@ -0,0 +1,132 @@
+	char stringTest1[22] = "1 2 + 9 * * 4 + 5 /";
+	int answerTest1 = 17;
+	printf("%s\n", stringTest1);
+	checkTest = testing(&headTest1, stringTest1, checkTest, answerTest1);
```
checkTest бывает либо 0, либо -1, я бы сделал его bool-ом

yurii-litvinov commented on HomeWork6/6.1/6.1/6.1/6.1.cpp:
```diff
@@ -0,0 +1,132 @@
+	char stringTest2[22] = "5 7 - 9 * 2 * 6 + 6 /";
+	int answerTest2 = -5;
+	printf("%s\n", stringTest2);
+	testing(&headTest2, stringTest2, checkTest, answerTest2);
```
Интересно, что тест 2 никак не влияет на checkTest :)

yurii-litvinov commented on HomeWork6/6.1/6.1/6.1/6.1.cpp:
```diff
@@ -0,0 +1,132 @@
+int main()
+{
+	tests();
+	return 0;
```
Оооокей. Напишите `return tests();` вместо этого. А когда переделаете tests на bool, напишите
```c++
return tests() ? 0 : -1;
```

yurii-litvinov reviewed (changes requested):
А  ещё надо сделать так, чтобы тесты ничего вообще на консоль не печатали,  и дать пользователю возможность свои значения ввести (после выполнения тестов, и только в случае, если тесты прошли успешно).

NikitaBabich26rus committed:
- Пользователь может вводить свои данные
- Testing сам создает стек
- checkTest теперь bool
- Еще исправлены мелкие недочеты

yurii-litvinov commented on HomeWork6/6.1/6.1/6.1/6.1.cpp:
```diff
@@ -0,0 +1,142 @@
+bool tests()
+{
+	bool checkTest1 = true;
+	char stringTest1[23] = "1 2 + 9 * 3 * 4 + 5 /\n";
```
А зачем '\n' в качестве стоп-символа, если каждая строка заканчивается всегда на '\0'?

yurii-litvinov commented on HomeWork6/6.1/6.1/6.1/6.1.cpp:
```diff
@@ -0,0 +1,142 @@
+	else
+	{
+		return false;
+	}
```
`return myAnswer == answer ? checkTest : false;`

yurii-litvinov commented on HomeWork6/6.1/6.1/6.1/6.1.cpp:
```diff
@@ -0,0 +1,142 @@
+	char stringTest1[23] = "1 2 + 9 * 3 * 4 + 5 /\n";
+	int answerTest1 = 17;
+	testCorrect(stringTest1);
+	int myAnswerTest1 = testing(stringTest1);
```
testCorrect же делается в testing, зачем делать его дважды?

yurii-litvinov commented on HomeWork6/6.1/6.1/6.1/6.1.cpp:
```diff
@@ -0,0 +1,142 @@
+	}
+}
+
+void checkString(Stack** head, char string[])
```
Очень странное название для функции, которая вычисляет значение выражения в постфиксной форме. И ещё более странный тип возвращаемого значения. Ведь если она считает, то почему бы ей не возвращать результат? И если так, то зачем ей стек передавать извне, она ведь сама могла бы его создать?

yurii-litvinov commented on HomeWork6/6.1/6.1/6.1/6.1.cpp:
```diff
@@ -0,0 +1,142 @@
+
+int testing(char stringTest[])
+{
+	if (testCorrect(stringTest))
```
Лучше 
```c++
if (!testCorrect(stringTest))
{
    return 0;
}

Stack* head = nullptr;
checkString(&head, stringTest);
int myAnswer = head->value;
deleteStack(&head);
return myAnswer;
```

yurii-litvinov commented on HomeWork6/6.1/6.1/6.1/6.1.cpp:
```diff
@@ -0,0 +1,142 @@
+	{
+		return -1;
+	}
+	else
```
Не надо этот else, если !tests(), мы уже выйдем

NikitaBabich26rus committed:
- Исправлены мелкие недочеты

yurii-litvinov reviewed (approved):
Окей, зачтена
